crates/nu-command/src/filesystem/du.rs

```diff
@@ -93,81 +96,69 @@ impl Command for Du {
-                    span: Span::unknown(),
-                },
-                &current_dir,
-                call.head,
-                None,
-            ),
-        }
-        .map(|f| f.1)?
-        .filter(move |p| {
-            if include_files {
-                true
-            } else {
-                matches!(p, Ok(f) if f.is_dir())
+        let paths = get_rest_for_glob_pattern(engine_state, stack, call, 0)?;
+        let paths = if paths.is_empty() { None } else { Some(paths) };
+
+        match paths {
+            None => {
+                let args = DuArgs {
+                    path: None,
+                    all,
+                    deref,
+                    exclude,
+                    max_depth,
+                    min_size,
+                };
+                Ok(
+                    du_for_one_pattern(args, &current_dir, tag, engine_state.ctrlc.clone())?
+                        .into_pipeline_data(engine_state.ctrlc.clone()),
+                )
             }
-        });
-
-        let all = args.all;
-        let deref = args.deref;
-        let max_depth = args.max_depth.map(|f| f.item as u64);
-        let min_size = args.min_size.map(|f| f.item as u64);
-
-        let params = DirBuilder {
-            tag,
-            min: min_size,
-            deref,
-            exclude,
-            all,
-        };
-
-        let mut output: Vec<Value> = vec![];
-        for p in paths.by_ref() {
-            match p {
-                Ok(a) => {
-                    if a.is_dir() {
-                        output.push(
-                            DirInfo::new(a, &params, max_depth, engine_state.ctrlc.clone()).into(),
-                        );
-                    } else if let Ok(v) = FileInfo::new(a, deref, tag) {
-                        output.push(v.into());
+            Some(paths) => {
+                let mut result_iters = vec![];
+                let mut errors = vec![];
+                for p in paths {
+                    let args = DuArgs {
+                        path: Some(p),
+                        all,
+                        deref,
+                        exclude: exclude.clone(),
+                        max_depth,
+                        min_size,
+                    };
+                    match du_for_one_pattern(args, &current_dir, tag, engine_state.ctrlc.clone()) {
+                        Ok(iter) => result_iters.push(iter),
+                        Err(e) => errors.push(e),
                     }
                 }
-                Err(e) => {
-                    output.push(Value::error(e, tag));
+                // if no results available, need to report all errors and return Err
+                // or else return Error along with items.
+                if result_iters.is_empty() {
+                    if !errors.is_empty() {
+                        let last_error = errors.pop().expect("Already check errors is not empty");
```

Comment:
```suggestion
                    if let Some(last_error) = errors.pop() {
```

Reply:
I don't think we need this again, because nushell will return error on partial error :)